app/src/main/java/org/mozilla/focus/utils/FirebaseHelper.java

```diff
@@ -79,223 +70,151 @@
+     * @param enable If true, enable Firebase Analytics.
+     */
+    public static void enableAnalytics(@Nullable final Context applicationContext, final boolean enable) {
 
-        return remoteConfigDefault;
-    }
+        checkFirebaseInitState();
+
+        // applicationContext is nullable for unit tests
+        if (applicationContext == null) {
+            return;
+        }
 
-    @Override
-    void refreshRemoteConfigDefault(Context context, RemoteConfigFetchCallback callback) {
-        // Clear remoteConfigDefault
-        remoteConfigDefault = null;
-        getRemoteConfigDefault(context);
-        // Now also need to reset the default config in Firebase if "Send Usage Data" is turned on.
-        enableRemoteConfig(context, TelemetryWrapper.isTelemetryEnabled(context), callback);
+        firebaseContract.enableAnalytics(applicationContext, enable);
     }
 
+    // all components required for app to function.
+    private static void initInternal(final Context applicationContext) {
 
-    private HashMap<String, Object> fromFile(Context context) {
+        checkFirebaseInitState();
 
-        // If we don't have read external storage permission, just don't bother reading the config file.
-        if (FileUtils.canReadExternalStorage(context)) {
-            try {
-                return FileUtils.fromJsonOnDisk(REMOTE_CONFIG_JSON);
-            } catch (Exception e) {
-                Log.w(TAG, "Some problem when reading RemoteConfig file from local disk: ", e);
-                // For any exception, we read the default resource file.
-                return fromResourceString(context);
-            }
+        // applicationContext is nullable for unit tests
+        if (applicationContext == null) {
+            return;
         }
+        firebaseContract.setDeveloperModeEnabled(AppConstants.isFirebaseBuild());
+        firebaseContract.init(applicationContext);
+        firebaseContract.enableRemoteConfig(applicationContext, () -> {
+            ThreadUtils.postToBackgroundThread(() -> {
+                final String pref = applicationContext.getString(R.string.pref_s_news);
+                final String source = firebaseContract.getRcString(pref);
+                final Settings settings = Settings.getInstance(applicationContext);
+                final boolean canOverride = settings.canOverride(PREF_INT_NEWS_PRIORITY, Settings.PRIORITY_FIREBASE);
+                Log.d(NewsSourceManager.TAG, "Remote Config fetched");
+                if (!TextUtils.isEmpty(source) && (canOverride || TextUtils.isEmpty(settings.getNewsSource()))) {
+                    Log.d(NewsSourceManager.TAG, "Remote Config is used:" + source);
+                    settings.setPriority(PREF_INT_NEWS_PRIORITY, Settings.PRIORITY_FIREBASE);
+                    settings.setNewsSource(source);
+                    NewsSourceManager.getInstance().setNewsSource(source);
```

Comment:
Just to leave a note, firebaseContract.enableRemoteConfig() only calls back if the config is fetched successfully. In case the operation is failed, the block here will not be executed, i.e. setNewSource() here will not be called. Since the method getNewsSource() will block until setNewsSource() is called, in another word, it's possible the caller of getNewsSource() will be blocked if the remote config is not fetched successfully.

Reply:
I'll fix it in https://github.com/mozilla-tw/FirefoxLite/issues/3472